### PR TITLE
Commenting of issues.

### LINKED ISSUE #685: Commenting of  issues [28h]
It will be possible to add any number of comment for an issue. Author of the comment will also be saved. The comments will be active across all runs.

Implementation
- [x] Add accordion panel to bug code view in GUI, list comments in time order, show author and date and text of comment, add text box and add button (8h)
- [x] Add API call to be able to add/query comment (based bughash:pathhash) (8h)
- [x] create new history table (author, date, text, bug hash)  (4h)
- [x] add tc (8h)
-----
28h

### PR DESCRIPTION
**Pull request by csordasmarton**
This commit resolves #685 issue.

### CODE REVIEW

File: tests/functional/comment/test_comment.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import os
+import unittest
+
+from codeCheckerDBAccess.ttypes import CommentData
+
+from libtest.thrift_client_to_db import get_all_run_results
+from libtest import env
+
+
+class TestSkeleton(unittest.TestCase):
```
Comment (whisperity) [inline tests/functional/comment/test_comment.py]: `class Test`**`Skeleton`**?

File: tests/functional/comment/test_comment.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+        # There are no comments available for the first bug
+        comments = self._cc_client.getComments(bug.bugHash)
+        self.assertEqual(len(comments), 0)
+
+        num_comment = self._cc_client.getCommentCount(bug.bugHash)
+        self.assertEqual(num_comment, 0)
+
+        # Try to add a new comment for the first bug
+        comment1 = CommentData(author = 'Anonymous', message = 'First msg')
```
Comment (whisperity) [inline tests/functional/comment/test_comment.py]: > pep8: No spaces between _kwargs_ keys and values.

File: www/scripts/codecheckerviewer/BugViewer.js [hunk 3]
```diff
@@ -7,6 +7,8 @@ define([
   'dojo/store/Memory',
   'dojo/store/Observable',
   'dojo/topic',
+  'dojo/fx',
```
Comment (whisperity) [inline www/scripts/codecheckerviewer/BugViewer.js]: Imports should be sorted in JS files too.

File: libcodechecker/orm_model.py [hunk 4]
```diff
@@ -267,6 +267,21 @@ def __init__(self, run_id, path, comment):
         self.run_id = run_id
         self.comment = comment
 
+class Comment(Base):
+    __tablename__ = 'comment'
```
Comment (csordasmarton) [inline libcodechecker/orm_model.py]: fixed

File: libcodechecker/orm_model.py [hunk 5]
```diff
@@ -267,6 +267,21 @@ def __init__(self, run_id, path, comment):
...
+
+    id = Column(Integer, autoincrement=True, primary_key=True)
+    hash = Column(String, nullable=False)
```
Comment (csordasmarton) [inline libcodechecker/orm_model.py]: fixed

File: tests/functional/comment/test_comment.py (hunk 1 above)
Comment (csordasmarton) [inline tests/functional/comment/test_comment.py]: fixed

File: tests/functional/comment/test_comment.py (hunk 2 above)
Comment (csordasmarton) [inline tests/functional/comment/test_comment.py]: fixed

File: www/scripts/codecheckerviewer/BugViewer.js (hunk 3 above)
Comment (csordasmarton) [inline www/scripts/codecheckerviewer/BugViewer.js]: fixed

File: libcodechecker/server/client_db_access_handler.py [hunk 6]
```diff
@@ -695,6 +696,145 @@ def getSuppressedBugs(self, run_id):
...
+                                              msg)
+
+    @timeit
+    def addComment(self, bug_hash, comment_data):
+        """
+            Add new comment for the given bug.
+        """
+        session = self.__session
+        comment = Comment(bug_hash,
+                          comment_data.author,
```
Comment (dkrupp) [inline libcodechecker/server/client_db_access_handler.py]: Security leak! 

Author must not be given by the client! It must be queried from the authenticated session. Wire in comment_data.author="Anonymous" for now and add a FIXME and a followup issue in GITHUB to query the proper user name.

Comment (whisperity) [inline libcodechecker/server/client_db_access_handler.py]: Also add a `# TODO: ` to this line, not _just_ a GitHub issue.

File: api/report_server.thrift [hunk 7]
```diff
@@ -204,6 +213,28 @@ service codeCheckerDBAccess {
...
+                              throws(1: shared.RequestFailed requestError),
+
+  // count all the comments for one bug
+  i64 getCommentCount(1: string bugHash)
+                      throws(1: shared.RequestFailed requestError),
+
+  // add new comment for a bug
+  bool addComment(1: string bugHash,
+                  2: CommentData comment)
+                 throws(1: shared.RequestFailed requestError),
```
Comment (whisperity) [inline api/report_server.thrift]: Indentation.

File: www/scripts/codecheckerviewer/BugViewer.js [hunk 8]
```diff
@@ -820,6 +826,47 @@ function (declare, dom, style, on, query, Memory, Observable, topic, entities,
...
+          });
+        },
+
+        onClick : function () {
+          if (this.hide) {
+            toggler.hide();
+            this.hide = false;
+          } else {
+            toggler.show();
+            this.hide = true
```
Comment (whisperity) [inline www/scripts/codecheckerviewer/BugViewer.js]: Missing `;`

File: www/scripts/codecheckerviewer/CommentView.js [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+
+      this._message = new ContentPane({
+        class   : 'message',
+        content : this.message
+      });
+
+      //--- Remove comment confirmation dialog ---//
+
+      this._removeDialog = new ConfirmDialog({
+        title     : 'CodeChecker - Remove comment',
```
Comment (whisperity) [inline www/scripts/codecheckerviewer/CommentView.js]: I think putting `CodeChecker` here is rendundant. The message is really short which already makes the box small.

File: www/scripts/codecheckerviewer/CommentView.js [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        hour   = minute * 60,
+        day    = hour * 24,
+        week   = day * 7,
+        month  = day * 30
+        year   = day * 365;
+
+    var fuzzy;
+
+    if (delta < 30) {
+      fuzzy = 'just then.';
```
Comment (whisperity) [inline www/scripts/codecheckerviewer/CommentView.js]: just **then**? When? `just now` works better. (Also, that's how GitHub and other such sites show it.)

**Pull request merged**